File: demo-task-executor-java/src/test/java/demo/executor/AsyncExecutorTaskOrderTest.java
package demo.executor;

import org.testng.annotations.Test;

import java.util.ArrayList;
import java.util.List;

import static org.testng.Assert.assertEquals;

/**
 * Unit test for {{@link demo.executor.AsyncExecutor}} to test run order of {@link demo.executor.Task}.
 */
@Test
public class AsyncExecutorTaskOrderTest extends AbstractAsyncExecutorTest {

    @Test
    public void testTaskOrder() throws InterruptedException {
        final int TASKS_COUNT = 8;

        List<Integer> consumer = new ArrayList<>();
        Executor executor = new AsyncExecutor(1);

        for (int i = 0; i < TASKS_COUNT; i++) {
            Task<?> task = getTestTask(consumer, i);
            executor.submit(task);
        }

        executor.waitUntilFinished();
        executor.stop();

        assertEquals(consumer.size(), TASKS_COUNT);
        for (int i = 0; i < TASKS_COUNT; i++) {
            assertEquals((int) consumer.get(i), i);
        }
    }
}

File: demo-task-executor-kotlin/src/test/kotlin/demo/executor/AsyncExecutorTaskOrderTest.kt
package demo.executor

import org.testng.Assert.assertEquals
import org.testng.Assert.fail
import org.testng.annotations.Test
import java.util.ArrayList

/**
 * Unit test for {[demo.executor.AsyncExecutor]} to test run order of [demo.executor.Task].
 */
@Test
class AsyncExecutorTaskOrderTest : AbstractAsyncExecutorTest() {

    @Test
    fun testTaskOrder() {
        val TASKS_COUNT = 8

        val consumer = ArrayList<Int>()
        val executor = AsyncExecutor(1)

        for (i in 0..TASKS_COUNT - 1) {
            val task = getTestTask(consumer, i)
            executor.submit(task)
        }

        executor.waitUntilFinished()
        executor.stop()

        assertEquals(consumer.size, TASKS_COUNT)
        for (i in 0..TASKS_COUNT - 1) {
            assertEquals(consumer[i], i)
        }
    }
}
